File: 2015/02/solution.js
let totalWrappingPaper = 0
let totalRibbon = 0

const processLine = (line) => {
  const sideLengths = parseLine(line)
  totalWrappingPaper += squareFeetNeeded(sideLengths)
  totalRibbon += ribbonNeeded(sideLengths)
}

const parseLine = (line) => {
  return line.split('x').map((sideLength) => parseInt(sideLength, 10))
}

const squareFeetNeeded = (sideLengths) => {
  // Since we're sorting, we know that the smallest side will be
  // the multiplication of the first two entries, so we can just
  // take three of those and call it a day instead of trying to do
  // any min-maxing.
  // Our formula then becomes 3x smallest side + 2x next side + 2x last side.
  sideLengths.sort((a, b) => a - b)
  return 3 * sideLengths[0] * sideLengths[1]
       + 2 * sideLengths[1] * sideLengths[2]
       + 2 * sideLengths[2] * sideLengths[0]
}

const ribbonNeeded = (sideLengths) => {
  // Like the wrapping paper solution, sorting makes this pretty simple.
  // We take 2x the shortest two sides plus the volume.
  sideLengths.sort((a, b) => a - b)
  return 2 * sideLengths[0]
       + 2 * sideLengths[1]
       + sideLengths[0] * sideLengths[1] * sideLengths[2]
}

const report = () => {
  console.log(`Part 1: The elves need ${totalWrappingPaper} square feet of wrapping paper.`)
  console.log(`Part 2: The elves need ${totalRibbon} feet of ribbon.`)
}

require('readline').
    createInterface({ input: require('fs').createReadStream('input.txt') }).
    on('line', processLine).
    on('close', report)

File: 2015/02/solution.rb
def square_feet_needed(sides)
  3 * sides[0] * sides[1] + 2 * sides[1] * sides[2] + 2 * sides[2] * sides[0]
end

def ribbon_needed(sides)
  2 * sides[0] + 2 * sides[1] + sides[0] * sides[1] * sides[2]
end

total_wrapping_paper = 0
total_ribbon = 0

File.open('input.txt') do |f|
  f.each_line do |line|
    side_lengths = line.split('x').map(&:to_i).sort
    total_wrapping_paper += square_feet_needed(side_lengths)
    total_ribbon += ribbon_needed(side_lengths)
  end
end

puts 'Part 1', total_wrapping_paper
puts 'Part 2' total_ribbon
